### PR TITLE
Optimise BlockStateEnum

### PR DESCRIPTION
**Pull request by AlfieC**
Partial solution to @oloflarsson's hopper issues.

### CODE REVIEW

File: Spigot-Server-Patches/0169-Cache-hashCode-in-BlockStateEnum.patch [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+
+diff --git a/src/main/java/net/minecraft/server/BlockStateEnum.java b/src/main/java/net/minecraft/server/BlockStateEnum.java
+index 71524f2..764160d 100644
+--- a/src/main/java/net/minecraft/server/BlockStateEnum.java
++++ b/src/main/java/net/minecraft/server/BlockStateEnum.java
+@@ -1,84 +1,76 @@
+ package net.minecraft.server;
+ 
+-import com.google.common.base.Optional;
```
Comment (kashike) [inline Spigot-Server-Patches/0169-Cache-hashCode-in-BlockStateEnum.patch]: Don't change imports.

File: Spigot-Server-Patches/0169-Cache-hashCode-in-BlockStateEnum.patch [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+-    }
+-
+-    public static <T extends Enum<T> & INamable> BlockStateEnum<T> of(String s, Class<T> oclass, T... at) {
+-        return a(s, oclass, (Collection) Lists.newArrayList(at));
+-    }
+-
+-    public static <T extends Enum<T> & INamable> BlockStateEnum<T> a(String s, Class<T> oclass, Collection<T> collection) {
+-        return new BlockStateEnum(s, oclass, collection);
++	
++  private final ImmutableSet<T> a;
```
Comment (kashike) [inline Spigot-Server-Patches/0169-Cache-hashCode-in-BlockStateEnum.patch]: What on earth are you doing? Fix your code style.

Comment (AlfieC) [inline Spigot-Server-Patches/0169-Cache-hashCode-in-BlockStateEnum.patch]: Wasn't my code style - was the .java file given to me by the import mc dev script. Will rerun and update.

**Comment by aikar** [comment]
Don't do the cacheing on demand. Do it immediately in the constructor as the last lines.

the hashCode method should be nothing but return hashCode;

the equals method also should be updated to check the hash codes before calling .equals on the set and map.

**Comment by AlfieC** [comment]
I tried doing it in the constructor but sometimes it would return a hash that was of no use. There is no disadvantage to doing it in the hashCode function - and since I'm not actually changing the hashCode there is no need to change equals.

**Comment by Spottedleaf** [comment]
@AlfieC 

> and since I'm not actually changing the hashCode there is no need to change equals.

aikar Is stating you can further this optimization by checking the hash code in equals before checking equality of the object. This is allowed because for any two equal objects their hashes must also be equal.

https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode()

> If two objects are equal according to the equals(Object) method, then calling the hashCode method on each of the two objects must produce the same integer result.

**Comment by Techcable** [comment]
Why not just use an [`EnumMap`](https://docs.oracle.com/javase/8/docs/api/java/util/EnumMap.html)? 
It's significantly more efficient than a regular map, and guava has builtin support for immutable enum maps with [`Maps.immutableEnumMap(Map<K extends Enum, V>)`](http://static.javadoc.io/com.google.guava/guava/17.0/com/google/common/collect/Maps.html#immutableEnumMap%28java.util.Map%29).

You can easily just turn `ImmutableMap.copyOf(map)` into `Maps.immutableEnumMap(map)` and fix your issue without changing the hash-code.

**Comment by aikar** [comment]
@Techcable  thats irrelevant to the change. This is an optimization that can improve anywhere the block states are compared.

It also becomes even more irrelevant when we get this moved to Arrays from your patch and not use the map at all.

**Comment by aikar** [comment]
@AlfieC  please put the hashmap calculation code in the constructor.

There is no reason to add the extra instructions (Which will cause harm towards inlining) to the hashCode method.

and yes as @Spottedleaf said. However speaking with Grum an whats going to happen in Vanilla, just replace the entire Equals method with object identity check (this == object)

**Comment by aikar** [comment]
also @Techcable, this is a case of bad naming. This class is actually not an enum.

**Comment by aikar** [comment]
@kashike pointed out you meant in the constructor of this, but the map uses enum as value not key, but could change the this.a to

this.a = Sets.immutableEnumSet(collection);

sure...

but that would be pointless. as the set is only iterated on in non hot code. Would be better off using a List, since its only exposed as a Collection<T> and not a Set.

File: Spigot-Server-Patches/0169-Optimise-hashCode-and-equals.patch [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+index 288c52c..c688d63 100644
+--- a/src/main/java/net/minecraft/server/BlockStateEnum.java
++++ b/src/main/java/net/minecraft/server/BlockStateEnum.java
+@@ -15,7 +15,12 @@ public class BlockStateEnum<T extends Enum<T> & INamable> extends BlockState<T>
+ 
+     private final ImmutableSet<T> a;
+     private final Map<String, T> b = Maps.newHashMap();
+-
++    
++    // Paper start - BlockStateEnum is a singleton, so we can use our own hashCode
```
Comment (kashike) [inline Spigot-Server-Patches/0169-Optimise-hashCode-and-equals.patch:17]: whitespace

**Comment by zachbr** [comment]
Merged w/ minor whitespace changes via d9afd08ceea0cddec8de06b5048f8b5193a4b5f3

**Pull request closed without merging**